spec/services/order_cycle_form_spec.rb

```diff
@@ -223,22 +223,93 @@
 
     context "updating payment methods" do
       context "and it's valid" do
-        it "saves the changes" do
-          distributor = create(:distributor_enterprise)
-          distributor_payment_method = create(
-            :payment_method,
-            distributors: [distributor]
-          ).distributor_payment_methods.first
-          order_cycle = create(:distributor_order_cycle, distributors: [distributor])
+        let!(:distributor){ create(:distributor_enterprise) }
+        let!(:payment_method){ create(:payment_method, distributors: [distributor]) }
+        let!(:payment_method2){ create(:payment_method, distributors: [distributor]) }
+        let!(:distributor_payment_method){ payment_method.distributor_payment_methods.first }
+        let!(:distributor_payment_method2){ payment_method2.distributor_payment_methods.first }
+        let!(:supplier){ create(:supplier_enterprise) }
+        context "the submitter is a coordinator" do
+          it "saves the changes" do
+            order_cycle = create(:distributor_order_cycle, distributors: [distributor])
+
+            form = OrderCycleForm.new(
+              order_cycle,
+              { selected_distributor_payment_method_ids: [distributor_payment_method.id] },
+              order_cycle.coordinator.users.first
+            )
 
-          form = OrderCycleForm.new(
-            order_cycle,
-            { selected_distributor_payment_method_ids: [distributor_payment_method.id] },
-            order_cycle.coordinator
-          )
+            expect(form.save).to be true
+            expect(order_cycle.distributor_payment_methods).to eq [distributor_payment_method]
+          end
+        end
 
-          expect(form.save).to be true
-          expect(order_cycle.distributor_payment_methods).to eq [distributor_payment_method]
+        context "submitter is a supplier" do
+          it "doesn't save the changes" do
+            order_cycle = create(:distributor_order_cycle, distributors: [distributor],
+                                                           suppliers: [supplier])
+
+            form = OrderCycleForm.new(
+              order_cycle,
+              { selected_distributor_payment_method_ids: [distributor_payment_method.id] },
+              supplier.users.first
+            )
+
+            expect(form).not_to receive(:attach_selected_distributor_payment_methods)
+            expect(order_cycle.distributor_payment_methods).to match_array [
+              distributor_payment_method, distributor_payment_method2
+            ]
```

Comment:
This is testing that the value didn't change, but the spec doesn't define the initial value. I think a better way is simply:

```suggestion
            expect{ form.save }.not_to change{ order_cycle.distributor_payment_methods }
```

Reply:
By the way, it looks like you forgot to call `form.save`, so I think this test would never fail! 
It's a good idea to make the spec fail to confirm that it actually tests the code right. Sometimes I comment-out part of the program code to break it and check that the spec fails. It's important to remember the order: red > green > refactor (but i still forget sometimes too).

Reply:
Good idea to use `change`.
I'll use it in all my test cases.